### PR TITLE
Durin adapter

### PR DESCRIPTION
**Pull request by cds95**
This PR implements the MVP for the Durin adapter.  It currently only supports a single endpoint namely `optimism-gateway`, which has the same functionality as the example gateway server implemented here https://github.com/smartcontractkit/ccip-read/tree/optimism-experimental/examples/optimism-gateway.

### CODE REVIEW

File: packages/sources/durin/README.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+The optimism global endpoint reads the latest proof from an Optimism as the L2 chain and returns the proof to the caller.
+Currently this endpoint has the same functionality as the server in this example https://github.com/smartcontractkit/ccip-read/tree/6d4deb917781f3becda39b9ebad6f21e037af1a6/examples/optimism-gateway.
+
+### Input Params
+
+| Required? |  Name  |                                              Description                                               | Options | Defaults to |
+| :-------: | :----: | :----------------------------------------------------------------------------------------------------: | :-----: | :---------: |
+|    ✅     | `abi`  | The ABI of the originally called L1 contract. In this example it is the OptimismResolverStub contract. |         |             |
+|    ✅     |  `to`  |                        The **L1** address of the originall called L1 contract.                         |         |
```
Comment (boxhock) [inline packages/sources/durin/README.md:33]:
```suggestion
|    ✅     |  `to`  |                        The **L1** address of the original called L1 contract.                         |         |
```

File: packages/sources/durin/schemas/env.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "$id": "https://external-adapters.chainlinklabs.com/schemas/durin.json",
+  "title": "@chainlink/durin env var schema",
+  "required": [],
```
Comment (boxhock) [inline packages/sources/durin/schemas/env.json:4]:
```suggestion
  "required": ["RPC_URL"],
```

File: packages/sources/durin/package.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "@chainlink/durin-adapter",
+  "version": "0.0.1",
```
Comment (boxhock) [inline packages/sources/durin/package.json:3]: We probably ought to update the example adapter, but I think we should start adapters on 1.0.0. @justinkaseman does this sound right?

```suggestion
  "version": "1.0.0",
```

File: packages/sources/durin/src/config.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+export interface Config extends BaseConfig {
+  l2RpcUrl?: string
+  addressManagerContract?: string
+}
+
+export const makeConfig = (prefix?: string): Config => {
+  const config = Requester.getDefaultConfig(prefix)
+  config.defaultEndpoint = DEFAULT_ENDPOINT
+  config.rpcUrl = util.getRequiredEnv('RPC_URL')
```
Comment (boxhock) [inline packages/sources/durin/src/config.ts:16]: Looks like this is unused

```suggestion
```

**Review by boxhock — CHANGES_REQUESTED** [review]
Do we have any integration tests for this adapter?
Reached out on Slack re: new release process

File: packages/sources/durin/package.json (hunk 3 above)
Comment (austinborn) [reply packages/sources/durin/package.json:3]: My vote would be 1.0.0 as well

File: packages/sources/durin/schemas/env.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  "type": "object",
+  "properties": {
+    "RPC_URL": {
+      "required": true,
```
Comment (austinborn) [inline packages/sources/durin/schemas/env.json:8]: Are these required fields used anywhere else? Long-term I would like to see a single source of truth, either these `required` attributes or the `required` array, but not both

File: packages/sources/durin/src/endpoint/optimism-gateway/index.ts [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  const hash = (el: Buffer | string): Buffer => {
+    return Buffer.from(ethers.utils.keccak256(el).slice(2), 'hex')
+  }
+  const leaves = elements.map((element) => {
+    return Buffer.from(element.slice(2), 'hex')
+  })
+  const tree = new MerkleTree(leaves, hash)
+  return tree.getProof(leaves[index], index).map((element) => {
+    return element.data
+  })
```
Comment (austinborn) [inline packages/sources/durin/src/endpoint/optimism-gateway/index.ts:171]:
```suggestion
  return tree.getProof(leaves[index], index).map(({ data }) => data)
```

File: packages/sources/durin/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { Fragment, Interface, JsonFragment } from '@ethersproject/abi'
+
+export const toInterface = (
+  abi: string | readonly (string | Fragment | JsonFragment)[] | Interface,
+) => {
+  if (Interface.isInterface(abi)) {
+    return abi
+  }
+  return new Interface(abi)
```
Comment (austinborn) [inline packages/sources/durin/src/utils.ts:9]:
```suggestion
  return Interface.isInterface(abi) ? abi : new Interface(abi)
```

File: packages/sources/durin/src/endpoint/optimism-gateway/index.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        stateRoots,
+      }
+    }
+  }
+  throw Error('No state root batches found')
+}
+
+const getElementsToConstructProof = (stateBatchHeader: StateBatchHeader): string[] => {
+  const elements = []
+  for (let i = 0; i < Math.pow(2, Math.ceil(Math.log2(stateBatchHeader.stateRoots.length))); i++) {
```
Comment (austinborn) [inline packages/sources/durin/src/endpoint/optimism-gateway/index.ts:151]:
```suggestion
  const maxElements = Math.pow(2, Math.ceil(Math.log2(stateBatchHeader.stateRoots.length)))
  for (let i = 0; i < maxElements; i++) {
```

**Review by austinborn — COMMENTED** [review]
Looks great, but I would like to see integration tests as well

**Comment by cds95** [comment]
@boxhock @austinborn I spoke to @thodges-gh and we are going to hold off on adding integration tests for this for now as it's still a MVP.  Currently the only test I've added runs makes a call to the original optimism-gateway server and compares it's output with the adapter's output.  I've created a ticket to add integration tests in the future.

https://app.shortcut.com/chainlinklabs/story/21369/add-integration-tests-to-durin-adapter

File: packages/sources/durin/schemas/env.json (hunk 5 above)
Comment (cds95) [reply packages/sources/durin/schemas/env.json:8]: I'm actually not sure about this.  @justinkaseman do you know the answer?

File: packages/sources/durin/test/unit/adapter.test.ts [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+import { assertError } from '@chainlink/ea-test-helpers'
+import { AdapterRequest } from '@chainlink/types'
+import { makeExecute } from '../../src/adapter'
+
+describe('execute', () => {
+  const jobID = '1'
+  const execute = makeExecute()
+  process.env.RPC_URL = 'http://localhost:9545'
+
+  describe('validation error', () => {
```
Comment (tateexon) [inline packages/sources/durin/test/unit/adapter.test.ts:11]: Looks like we only have tests for validation errors. Can we add a test that would pass validation and return a 200?

Comment (cds95) [reply packages/sources/durin/test/unit/adapter.test.ts:11]: I don't think any of our adapters have unit tests that pass validation and return a 200 right now.  These are typically done in the integration tests which we are holding off for now.

**Review by austinborn — APPROVED** [review]

**Review by boxhock — APPROVED** [review]

**Pull request merged**